Typing deep matrices (tensors)

Recently, we wanted to better type certain cases. Since the `algebra` lib can handle arrays dimensions from 1...N, we needed to better specify types for Matrix and ND Tensors greater than 2 dimensions.

Our thought was that we could get away with the following using forward referencing:

```py
# Vectors, Matrices, and Arrays are assumed to be mutable lists
Vector = List[float]
Matrix = List[Vector]
Tensor = List[Union[Matrix, 'Tensor']]
Array = Union[Matrix, Vector, Tensor]
```

I ended up exploring this a bit more, and while `mypy` will accept this and seem to function normally, in reality, it doesn't know what the heck is going on.

<details>
<summary>Test Example</summary>

```py
"""Typing."""
from __future__ import annotations
from typing import Union, Sequence, List, Any

# Vectors, Matrices, and Arrays are assumed to be mutable lists
Vector = List[float]
Matrix = List[Vector]
Tensor = List[List[List[Any]]]
Array = Union[Matrix, Vector, Tensor]

# Anything that resembles a sequence will be considered "like" one of our types above
VectorLike = Sequence[float]
MatrixLike = Sequence[VectorLike]
TensorLike = Sequence[Sequence[Sequence[Any]]]
ArrayLike = Union[VectorLike, MatrixLike, TensorLike]


def func1(a: Tensor):
    print(a)


func1([[[3.0, 3.0]]])

func1([[[[3.0, 3.0]]]])

func1([[[[[3.0, 3.0]]]]])


def func2(a: Matrix):
    print(a)


func2([[3, 3]])


def func3(a: Vector):
    print(a)


func3([3, 3])
```

</details>

```console
➜  typecheck python3 -m mypy test.py
test.py:24: error: List item 0 has incompatible type "list[list[list[float]]]"; expected "list[list[float]] | Tensor"  [list-item]
test.py:26: error: List item 0 has incompatible type "list[list[list[list[float]]]]"; expected "list[list[float]] | Tensor"  [list-item]
```

The reality is that we simply cannot do recursive types. There is no real support for this, even if it appears it might know what it is looking at.

The best we can probably do is something like this:

```py
# Vectors, Matrices, and Arrays are assumed to be mutable lists
Vector = List[float]
Matrix = List[Vector]
Tensor = List[List[List[Any]]]
Array = Union[Matrix, Vector, Tensor]
```

This will keep basic typing working for vectors and matrices. Anything beyond 2 dimensions will fall in the more permissive `Tensor` type which just considers Any after the 3rd dimension.

```console
➜  typecheck python3 -m mypy test.py
Success: no issues found in 1 source file
```

This isn't urgent, we don't use tensors anywhere in the code, right now it is more to separate tensors from simple matrices (2D tensors). We should probably use this new approach moving forward that what we have now which currently passes testing, but in reality will never actually work.

--- a/coloraide/algebra.py
+++ b/coloraide/algebra.py
@@ -2638,7 +2638,7 @@ def lu(
             size = s[1]
             wide = True
             for _ in range(diff):
-                matrix.append([0.0] * size)  # type: ignore[arg-type]  # noqa: PERF401
+                matrix.append([0.0] * size)  # type: ignore[list-item]  # noqa: PERF401
         # Tall
         else:
             tall = True

--- a/coloraide/algebra.py
+++ b/coloraide/algebra.py
@@ -686,13 +686,13 @@ def pprint(value: float | ArrayLike) -> None:
 def all(a: float | ArrayLike) -> bool:  # noqa: A001
     """Return true if all elements are "true"."""
 
-    return _all(flatiter(a))  # type: ignore[arg-type]
+    return _all(flatiter(a))
 
 
 def any(a: float | ArrayLike) -> bool:  # noqa: A001
     """Return true if all elements are "true"."""
 
-    return _any(flatiter(a))  # type: ignore[arg-type]
+    return _any(flatiter(a))
 
 
 def vdot(a: VectorLike, b: VectorLike) -> float:
@@ -1272,7 +1272,7 @@ def multi_dot(arrays: Sequence[ArrayLike]) -> Any:
             is_vector = True
 
     # Make sure everything is a 2-D matrix as the next calculations only work for 2-D.
-    if not _all(len(s) == 2 for s in shapes):  # type: ignore[arg-type]
+    if not _all(len(s) == 2 for s in shapes):
         raise ValueError('All arrays must be 2-D matrices')
 
     # No need to do the expensive and complicated chain order algorithm for only 3.

--- a/coloraide/types.py
+++ b/coloraide/types.py
@@ -10,24 +10,24 @@ ColorInput = Union['Color', str, Mapping[str, Any]]
 # Vectors, Matrices, and Arrays are assumed to be mutable lists
 Vector = List[float]
 Matrix = List[Vector]
-Tensor = List[List[List[Any]]]
+Tensor = List[List[List[Union[float, Any]]]]
 Array = Union[Matrix, Vector, Tensor]
 
 # Anything that resembles a sequence will be considered "like" one of our types above
 VectorLike = Sequence[float]
 MatrixLike = Sequence[VectorLike]
-TensorLike = Sequence[Sequence[Sequence[Any]]]
+TensorLike = Sequence[Sequence[Sequence[Union[float, Any]]]]
 ArrayLike = Union[VectorLike, MatrixLike, TensorLike]
 
 # Vectors, Matrices, and Arrays of various, specific types
 VectorBool = List[bool]
 MatrixBool = List[VectorBool]
-TensorBool = List[Union[MatrixBool, 'TensorBool']]
+TensorBool = List[List[List[Union[bool, Any]]]]
 ArrayBool = Union[MatrixBool, VectorBool, TensorBool]
 
 VectorInt = List[int]
 MatrixInt = List[VectorInt]
-TensorInt = List[Union[MatrixInt, 'TensorInt']]
+TensorInt = List[List[List[Union[int, Any]]]]
 ArrayInt = Union[MatrixInt, VectorInt, TensorInt]
 
 # General algebra types

--- a/coloraide/types.py
+++ b/coloraide/types.py
@@ -10,13 +10,13 @@ ColorInput = Union['Color', str, Mapping[str, Any]]
 # Vectors, Matrices, and Arrays are assumed to be mutable lists
 Vector = List[float]
 Matrix = List[Vector]
-Tensor = List[Union[Matrix, 'Tensor']]
+Tensor = List[List[List[Any]]]
 Array = Union[Matrix, Vector, Tensor]
 
 # Anything that resembles a sequence will be considered "like" one of our types above
 VectorLike = Sequence[float]
 MatrixLike = Sequence[VectorLike]
-TensorLike = Sequence[Union[MatrixLike, 'TensorLike']]
+TensorLike = Sequence[Sequence[Sequence[Any]]]
 ArrayLike = Union[VectorLike, MatrixLike, TensorLike]
 
 # Vectors, Matrices, and Arrays of various, specific types

--- a/docs/src/markdown/about/changelog.md
+++ b/docs/src/markdown/about/changelog.md
@@ -1,5 +1,9 @@
 # Changelog
 
+## Unreleased
+
+-   **FIX**: Fix typing of deeply nested arrays in `algebra`.
+
 ## 2.13.1
 
 -   **FIX**: Minor typing regressions and fixes.

